packages/core/src/wot-impl.ts

```diff
@@ -22,33 +22,34 @@ import Helpers from "./helpers";
 import { createLoggers } from "./logger";
 import ContentManager from "./content-serdes";
 import { getLastValidationErrors, isThingDescription } from "./validation";
+import { inspect } from "util";
 
 const { debug } = createLoggers("core", "wot-impl");
 
 class ThingDiscoveryProcess implements WoT.ThingDiscoveryProcess {
-    constructor(rawThingDescriptions: WoT.DataSchemaValue, filter?: WoT.ThingFilter) {
+    constructor(private directory: ConsumedThing, public filter?: WoT.ThingFilter) {
         this.filter = filter;
         this.done = false;
-        this.rawThingDescriptions = rawThingDescriptions;
     }
 
-    rawThingDescriptions: WoT.DataSchemaValue;
-
-    filter?: WoT.ThingFilter | undefined;
     done: boolean;
     error?: Error | undefined;
     async stop(): Promise<void> {
         this.done = true;
     }
 
     async *[Symbol.asyncIterator](): AsyncIterator<WoT.ThingDescription> {
-        if (!(this.rawThingDescriptions instanceof Array)) {
-            this.error = new Error("Encountered an invalid output value.");
+        let rawThingDescriptions: WoT.ThingDescription[];
+        try {
+            const thingsPropertyOutput = await this.directory.readProperty("things");
+            rawThingDescriptions = (await thingsPropertyOutput.value()) as WoT.ThingDescription[];
```

Comment:
Could we use the generic "version" of `value<T>` here to avoid the casting?

```suggestion
            rawThingDescriptions = await thingsPropertyOutput.value<WoT.ThingDescription[]>());
```

Reply:
Sadly no, Given the fact that CosumeThing implementation returns `WoT.InteractionOutput` and not our `InteractionOutput`. As I said another time the generic in the value function is mostly experimental... we should think if it makes sense to have it in the public APIs.

Reply:
I see, then we can mark this thread as resolved :) Or should we open an issue for the evaluation of the generic in the value function?

Reply:
I think the issue should be raised in the Scripting API repo, first (maybe it is already there?) and if it fails we have to evaluate how to do it here. @danielpeintner what do you think?

Reply:
I am not aware of a related issue in the Scripting API repo. Raising the aspect makes sense and maybe gives some more feedback from others...